sdk/eventhub/Azure.Messaging.EventHubs/src/EventProcessorClient.cs

```diff
@@ -287,194 +437,404 @@ public class EventProcessorClient : EventProcessorBase<PartitionContext>, IAsync
-                catch (Exception)
+                finally
                 {
-                    // TODO: delegate the exception handling to an Exception Callback.
-                    // Maybe we should just surface the exception.
+                    RunningTaskSemaphore.Release();
                 }
             }
         }
 
         /// <summary>
-        ///   Responsible for processing events received from the Event Hubs service.
+        ///   Stops the event processor.  In case it isn't running, nothing happens.
         /// </summary>
         ///
-        /// <param name="partitionEvent">The partition event to be processed.</param>
-        /// <param name="context">The context in which the associated partition is being processed.</param>
+        /// <param name="cancellationToken">A <see cref="CancellationToken"/> instance to signal the request to cancel the operation.</param>
+        ///
+        public virtual void StopProcessing(CancellationToken cancellationToken = default) => StopProcessingAsync(cancellationToken).GetAwaiter().GetResult();
+
+        /// <summary>
+        ///   Determines whether the specified <see cref="System.Object" /> is equal to this instance.
+        /// </summary>
+        ///
+        /// <param name="obj">The <see cref="System.Object" /> to compare with this instance.</param>
+        ///
+        /// <returns><c>true</c> if the specified <see cref="System.Object" /> is equal to this instance; otherwise, <c>false</c>.</returns>
+        ///
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public override bool Equals(object obj) => base.Equals(obj);
+
+        /// <summary>
+        ///   Returns a hash code for this instance.
+        /// </summary>
+        ///
+        /// <returns>A hash code for this instance, suitable for use in hashing algorithms and data structures like a hash table.</returns>
+        ///
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public override int GetHashCode() => base.GetHashCode();
+
+        /// <summary>
+        ///   Converts the instance to string representation.
+        /// </summary>
+        ///
+        /// <returns>A <see cref="System.String" /> that represents this instance.</returns>
+        ///
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public override string ToString() => base.ToString();
+
+        /// <summary>
+        ///   Updates the checkpoint using the given information for the associated partition and consumer group in the chosen storage service.
+        /// </summary>
+        ///
+        /// <param name="eventData">The event containing the information to be stored in the checkpoint.</param>
+        /// <param name="context">The context of the partition the checkpoint is associated with.</param>
+        ///
+        /// <returns>A task to be resolved on when the operation has completed.</returns>
+        ///
+        internal Task InternalUpdateCheckpointAsync(EventData eventData,
```

Comment:
What's the scenario for this one?  Seems to be taking the same parameters as what its calling.

Reply:
It was created because of the base class. `UpdateCheckpointAsync` was `protected` only (and we wanted to keep it like that in the base class). However, `ProcessEventArgs` needs to call this method to update checkpoints. For this reason we created this `internal` method.

I changed `UpdateCheckpointAsync` to `private` while we don't know what approach we will take for the base class. I could make it `internal` only if you think it's better this way.

Reply:
It's all implementation, so it doesn't really matter... but we should probably collapse down to the single method marked `internal` at some point.